src/clusterfuzz/_internal/system/archive.py

```diff
@@ -57,88 +247,38 @@ def iterator(archive_path,
-
-    def file_match_callback_fallback(_):
-      return True
-
-    file_match_callback = file_match_callback_fallback
-
-  def maybe_extract(extract_func, info):
-    """Returns an extracted file or None if it is not supposed to be extracted.
-    """
-    if should_extract:
-      return extract_func(info)
-    return None
-
-  if archive_type == ArchiveType.ZIP:
-    try:
-      with zipfile.ZipFile(archive_obj or archive_path) as zip_file:
-        for info in zip_file.infolist():
-          if not file_match_callback(info.filename):
-            continue
-
-          yield ArchiveFile(info.filename, info.file_size,
-                            maybe_extract(zip_file.open, info))
-
-    except (zipfile.BadZipfile, zipfile.LargeZipFile):
-      logs.log_error('Bad zip file %s.' % archive_path)
-
-  elif archive_type == ArchiveType.TAR:
-    try:
-      if archive_obj:
-        tar_file = tarfile.open(fileobj=archive_obj)
-      else:
-        tar_file = tarfile.open(archive_path)
-
-      for info in tar_file.getmembers():
-        if not file_match_callback(info.name):
-          continue
-
-        yield ArchiveFile(info.name, info.size,
-                          maybe_extract(tar_file.extractfile, info))
-      tar_file.close()
-    except tarfile.TarError:
-      logs.log_error('Bad tar file %s.' % archive_path)
-
-  elif archive_type == ArchiveType.TAR_LZMA:
-    assert archive_obj is None, "LZMAFile doesn't support opening file handles."
-    try:
-      with lzma.LZMAFile(archive_path) as lzma_file, \
-            tarfile.open(fileobj=lzma_file) as tar_file:
-
-        error_filepaths = []
-        for info in tar_file.getmembers():
-          if not file_match_callback(info.name):
-            continue
-
-          try:
-            yield ArchiveFile(info.name, info.size,
-                              maybe_extract(tar_file.extractfile, info))
-
-          except KeyError:  # Handle broken links gracefully.
-            error_filepaths.append(info.name)
-            yield ArchiveFile(info.name, info.size, None)
-
-        if error_filepaths:
-          logs.log_warn(
-              'Check archive %s for broken links.' % archive_path,
-              error_filepaths=error_filepaths)
-
-    except (lzma.LZMAError, tarfile.TarError):
-      logs.log_error('Bad lzma file %s.' % archive_path)
-
-  else:
-    logs.log_error('Unsupported compression type for file %s.' % archive_path)
-
-
-def extracted_size(archive_path, archive_obj=None, file_match_callback=None):
+  try:
+    reader = get_archive_reader(archive_path, archive_obj)
+    assert reader is not None
+
+    files = [
+        f for f in reader.list_files()
+        if not file_match_callback or file_match_callback(f.filename)
+    ]
+    error_file_paths = []
+    for file in files:
+      try:
+        yield ArchiveFile(
+            file.filename,
+            file.file_size,
+            handle=None if not should_extract else reader.open(file.filename))
```

Comment:
`reader.open()` is the only operation that can throw here, let's rejigger the code to make this explicit:

```py
handle = None
try:
  handle = reader.open(file.filename)
except KeyError:
  error_file_paths.append(file.filename)

yield ArchiveFile(file.filename, file.file_size, handle)
```